### PR TITLE
defensive write and undo

### PR DESCRIPTION
**Pull request by murali-shris**
**- What I did**

handle exception on socket write and provide framework for undo when we are unable to write back to client app

**- How I did it**

introduce undo in abstract verb handler. 

**- How to verify it**
#TODO

**- Description for the changelog**
# TODO.

### CODE REVIEW

File: at_secondary/at_secondary_server/lib/src/exception/global_exception_handler.dart [hunk 1]
```diff
@@ -121,6 +121,10 @@ class GlobalExceptionHandler {
 
   Future<void> _writeToSocket(AtConnection atConnection, String prompt,
       String error_code, String error_description) async {
-    atConnection.write('error:$error_code-$error_description\n$prompt');
+    try {
+      atConnection.write('error:$error_code-$error_description\n$prompt');
+    } on ConnectionInvalidException {
+      // do Nothing. Unable to write error code back to client
```
Comment (VJag) [inline at_secondary/at_secondary_server/lib/src/exception/global_exception_handler.dart]: Lets log this

File: at_secondary/at_secondary_server/lib/src/connection/inbound/inbound_message_listener.dart [hunk 2]
```diff
@@ -76,6 +76,7 @@ class InboundMessageListener {
   }
 
   Future<void> _closeConnection() async {
+    _buffer.clear();
```
Comment (VJag) [inline at_secondary/at_secondary_server/lib/src/connection/inbound/inbound_message_listener.dart:79]: This should be done on close and on clear and on request completion

File: at_secondary/at_secondary_server/lib/src/verb/handler/abstract_verb_handler.dart [hunk 3]
```diff
@@ -32,7 +32,15 @@ abstract class AbstractVerbHandler implements VerbHandler {
   Future<void> process(String command, InboundConnection atConnection) async {
     var response = await processInternal(command, atConnection);
     var handler = responseManager.getResponseHandler(getVerb());
-    handler.process(atConnection, response);
+    try {
+      handler.process(atConnection, response);
+    } on Exception {
+      if (handler.isComplete()) {
```
Comment (VJag) [inline at_secondary/at_secondary_server/lib/src/verb/handler/abstract_verb_handler.dart]: // I think we can hold isComplete() for now. For two reasons:

1. This is request scoped
2. Calling handler.isComplete not in the context of process does not add value

My 2 cents:

try {
      handler.process(atConnection, response);
    } on Exception {
      // Undo needs command as input
      await undo(command);
    }

File: at_secondary/at_secondary_server/lib/src/verb/handler/response/base_response_handler.dart [hunk 4]
```diff
@@ -34,7 +34,11 @@ abstract class BaseResponseHandler implements ResponseHandler {
           ? '$atSign@'
           : (isPolAuthenticated ? '$fromAtSign@' : '@');
       var responseMessage = getResponseMessage(result, prompt);
-      connection.write(responseMessage);
+      try {
+        connection.write(responseMessage);
+      } on ConnectionInvalidException {
+        rethrow;
```
Comment (VJag) [inline at_secondary/at_secondary_server/lib/src/verb/handler/response/base_response_handler.dart]: We need not catch as we are not handling this

File: at_secondary/at_secondary_server/lib/src/connection/inbound/inbound_message_listener.dart (hunk 2 above)
Comment (murali-shris) [reply at_secondary/at_secondary_server/lib/src/connection/inbound/inbound_message_listener.dart:79]: buffer clear is done in _messageHandler for other scenarios(exception,end of request etc.,)

File: at_secondary/at_secondary_server/lib/src/verb/handler/abstract_verb_handler.dart [hunk 5]
```diff
@@ -55,6 +59,11 @@ abstract class AbstractVerbHandler implements VerbHandler {
     }
   }
 
+  //# TODO move undo to spec
+  Future<void> undo() async {
```
Comment (VJag) [inline at_secondary/at_secondary_server/lib/src/verb/handler/abstract_verb_handler.dart]: Undo should take command as an argument

**Review by VJag — APPROVED** [review]

**Comment by murali-shris** [comment]
changes ported to sdk_refactoring branch.

**Pull request closed without merging**